### PR TITLE
Initial Commit of System.Threading.Overlapped

### LINKED ISSUE #906: Open Source System.Threading.Overlapped.dll

### PR DESCRIPTION
**Pull request by FiveTimesTheFun**
Fixes #906

### CODE REVIEW

File: src/System.Threading.Overlapped/src/System.Threading.Overlapped.CoreCLR.csproj [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <PropertyGroup>
+    <Configuration Condition=" '$(Configuration)' == '' ">Windows_Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <AssemblyVersion>4.0.0.0</AssemblyVersion>
+    <AssemblyName>System.Threading.Overlapped</AssemblyName>
+    <OutputType>Library</OutputType>
+    <ProjectGuid>{6A07CCB8-3E59-47e7-B3DD-DB1F6FC501D5}</ProjectGuid>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <ResInputPath>$(MSBuildThisFileDirectory)Resources</ResInputPath>
```
Comment (mellinoe) [inline src/System.Threading.Overlapped/src/System.Threading.Overlapped.CoreCLR.csproj]: nit: I think this is the default, right?

File: src/System.Threading.Overlapped/src/System.Threading.Overlapped.CoreCLR.csproj [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    <IsPartialFacadeAssembly>true</IsPartialFacadeAssembly>
+    <NuGetTargetFrameworkMoniker>DNXCore,Version=v5.0</NuGetTargetFrameworkMoniker>    
+  </PropertyGroup>
+  <!-- Help VS understand available configurations -->
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Windows_Debug|AnyCPU' " />
```
Comment (mellinoe) [inline src/System.Threading.Overlapped/src/System.Threading.Overlapped.CoreCLR.csproj]: Is there any platform-specific code for these? Or will that part be limited to the types in mscorlib?

Comment (stephentoub) [inline src/System.Threading.Overlapped/src/System.Threading.Overlapped.CoreCLR.csproj]: This whole contract should be considered Windows only I believe.

File: src/System.Threading.Overlapped/src/System/Threading/ClrThreadPoolBoundHandle.cs [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+        }
+
+        private static unsafe ThreadPoolBoundHandleOverlapped GetOverlappedWrapper(NativeOverlapped* overlapped, ThreadPoolBoundHandle expectedBoundHandle)
+        {
+            ThreadPoolBoundHandleOverlapped wrapper;
+            try
+            {
+                wrapper = (ThreadPoolBoundHandleOverlapped)Overlapped.Unpack(overlapped);
+            }
+            catch (NullReferenceException ex)
```
Comment (stephentoub) [inline src/System.Threading.Overlapped/src/System/Threading/ClrThreadPoolBoundHandle.cs]: Ugh

File: src/System.Threading.Overlapped/tests/System.Threading.Overlapped.Tests.csproj [hunk 4]
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <PropertyGroup>
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <OutputType>Library</OutputType>
+    <AssemblyName>System.Threading.Overlapped.Tests</AssemblyName>
+    <ProjectGuid>{861A3318-35AD-46ac-8257-8D5D2479BAD9}</ProjectGuid>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Comment (stephentoub) [inline src/System.Threading.Overlapped/tests/System.Threading.Overlapped.Tests.csproj]: This project should be marked as not being usable on Linux or OSX, e.g. https://github.com/dotnet/corefx/blob/master/src/Microsoft.Win32.Registry/tests/Microsoft.Win32.Registry.Tests.csproj#L11

File: src/System.Threading.Overlapped/tests/ThreadPoolBoundHandle_AllocateNativeOverlappedTests.cs [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+using System.Runtime.InteropServices;
+using System.Threading;
+using Xunit;
+
+public partial class ThreadPoolBoundHandleTests
+{
+    [Fact]
+    public unsafe void AllocateNativeOverlapped_NullAsCallback_ThrowsArgumentNullException()
+    {
+        ThreadPoolBoundHandle handle = CreateThreadPoolBoundHandle();
```
Comment (stephentoub) [inline src/System.Threading.Overlapped/tests/ThreadPoolBoundHandle_AllocateNativeOverlappedTests.cs]: I know these are "just tests", but should we be disposing of things like this?

File: src/System.Threading.Overlapped/tests/ThreadPoolBoundHandle_GetNativeOverlappedStateTests.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+        object result = ThreadPoolBoundHandle.GetNativeOverlappedState(overlapped);
+
+        Assert.Same(context, result);
+    }
+
+    [Fact]
+    public unsafe void GetNativeOverlappedState_WhenUnderlyingStateIsIAsyncResult_ReturnsIAsyncResult()
+    {   // CoreCLR/Desktop CLR version of overlapped sits on top of Overlapped class 
+        // and treats IAsyncResult specially, which is why we special case this case.
```
Comment (stephentoub) [inline src/System.Threading.Overlapped/tests/ThreadPoolBoundHandle_GetNativeOverlappedStateTests.cs]: @ericeil, is this still true?

File: src/System.Threading.Overlapped/src/System/Threading/DeferredDisposableLifetime.cs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+                    // We haven't been disposed.  Decrement _count.
+                    int newCount = oldCount - 1;
+                    if (Interlocked.CompareExchange(ref _count, newCount, oldCount) == oldCount)
+                    {
+                        if (newCount == 0)
+                            obj.OnFinalRelease(disposed: false);
+                        return;
+                    }
+                }
+                else
```
Comment (stephentoub) [inline src/System.Threading.Overlapped/src/System/Threading/DeferredDisposableLifetime.cs]: Just took a quick look at code coverage... looks like no tests hit this whole else block for the case of oldCount <= 0.

File: src/System.Threading.Overlapped/src/System/Threading/ClrThreadPoolBoundHandle.cs [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+
+            EnsureNotDisposed();
+
+            preAllocated.AddRef();
+            try
+            {
+                ThreadPoolBoundHandleOverlapped overlapped = preAllocated._overlapped;
+
+                if (overlapped._boundHandle != null)
+                    throw new ArgumentException(SR.Argument_PreAllocatedAlreadyAllocated, "preAllocated");
```
Comment (stephentoub) [inline src/System.Threading.Overlapped/src/System/Threading/ClrThreadPoolBoundHandle.cs]: Is it possible to add a test that triggers/verifies this exception?

File: src/System.Threading.Overlapped/src/System/Threading/ClrThreadPoolBoundHandle.cs [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+        {
+            if (overlapped == null)
+                throw new ArgumentNullException("overlapped");
+
+            // Note: we explicitly allow FreeNativeOverlapped calls after the ThreadPoolBoundHandle has been Disposed.
+
+            ThreadPoolBoundHandleOverlapped wrapper = GetOverlappedWrapper(overlapped, this);
+
+            if (wrapper._boundHandle != this)
+                throw new ArgumentException(SR.Argument_NativeOverlappedWrongBoundHandle, "overlapped");
```
Comment (stephentoub) [inline src/System.Threading.Overlapped/src/System/Threading/ClrThreadPoolBoundHandle.cs]: Ditto on a test.

**Comment by stephentoub** [comment]
A few minor things to be addressed, but otherwise LGTM.

**Comment by stephentoub** [comment]
Four of the tests failed in both the debug and release CI runs:
```
Discovering: System.Threading.Overlapped.Tests
  Discovered:  System.Threading.Overlapped.Tests
  Starting:    System.Threading.Overlapped.Tests
     ThreadPoolBoundHandleTests.SingleOperationOverSingleHandle [FAIL]
        Assert.Equal() Failure
        Expected: 1
        Actual:   0
        Stack Trace:
           d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Threading.Overlapped\tests\ThreadPoolBoundHandle_IntegrationTests.cs(31,0): at ThreadPoolBoundHandleTests.SingleOperationOverSingleHandle()
     ThreadPoolBoundHandleTests.MultipleOperationsOverSingleHandle [FAIL]
        Assert.Equal() Failure
        Expected: 1
        Actual:   0
        Stack Trace:
           d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Threading.Overlapped\tests\ThreadPoolBoundHandle_IntegrationTests.cs(73,0): at ThreadPoolBoundHandleTests.MultipleOperationsOverSingleHandle()
     ThreadPoolBoundHandleTests.FlowsAsyncLocalsToCallback [FAIL]
        Assert.Equal() Failure
        Expected: 1
        Actual:   0
        Stack Trace:
           d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Threading.Overlapped\tests\ThreadPoolBoundHandle_IntegrationTests.cs(195,0): at ThreadPoolBoundHandleTests.FlowsAsyncLocalsToCallback()
     ThreadPoolBoundHandleTests.MultipleOperationsOverMultipleHandles [FAIL]
        Assert.Equal() Failure
        Expected: 1
        Actual:   0
        Stack Trace:
           d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.Threading.Overlapped\tests\ThreadPoolBoundHandle_IntegrationTests.cs(133,0): at ThreadPoolBoundHandleTests.MultipleOperationsOverMultipleHandles()
  Finished:    System.Threading.Overlapped.Tests
```

File: src/System.Threading.Overlapped/src/System/Threading/ClrThreadPoolBoundHandle.cs (hunk 3 above)
Comment (FiveTimesTheFun) [inline src/System.Threading.Overlapped/src/System/Threading/ClrThreadPoolBoundHandle.cs]: At least it's not normal code path, just wraps and throws a new exception type.

Do you know why they didn't compare overlapped to null at the beginning of this method?

File: src/System.Threading.Overlapped/src/System.Threading.Overlapped.CoreCLR.csproj (hunk 1 above)
Comment (FiveTimesTheFun) [inline src/System.Threading.Overlapped/src/System.Threading.Overlapped.CoreCLR.csproj]: Thanks, updated.

File: src/System.Threading.Overlapped/src/System.Threading.Overlapped.CoreCLR.csproj (hunk 2 above)
Comment (FiveTimesTheFun) [inline src/System.Threading.Overlapped/src/System.Threading.Overlapped.CoreCLR.csproj]: Thanks, updated.

File: src/System.Threading.Overlapped/tests/System.Threading.Overlapped.Tests.csproj (hunk 4 above)
Comment (FiveTimesTheFun) [inline src/System.Threading.Overlapped/tests/System.Threading.Overlapped.Tests.csproj]: Thanks, updated.

File: src/System.Threading.Overlapped/tests/ThreadPoolBoundHandle_AllocateNativeOverlappedTests.cs (hunk 5 above)
Comment (FiveTimesTheFun) [inline src/System.Threading.Overlapped/tests/ThreadPoolBoundHandle_AllocateNativeOverlappedTests.cs]: Thanks, I've done this for all of the tests.

File: src/System.Threading.Overlapped/src/System/Threading/ClrThreadPoolBoundHandle.cs (hunk 8 above)
Comment (FiveTimesTheFun) [inline src/System.Threading.Overlapped/src/System/Threading/ClrThreadPoolBoundHandle.cs]: Thanks, updated.

File: src/System.Threading.Overlapped/src/System/Threading/ClrThreadPoolBoundHandle.cs (hunk 9 above)
Comment (FiveTimesTheFun) [inline src/System.Threading.Overlapped/src/System/Threading/ClrThreadPoolBoundHandle.cs]: Thanks, added FreeNativeOverlapped_WithWrongHandle_ThrowsArgumentException

**Comment by Maxwe11** [comment]
@dotnet-bot test this please

File: src/System.Threading.Overlapped/tests/ThreadPoolBoundHandle_GetNativeOverlappedStateTests.cs (hunk 6 above)
Comment (ericeil) [inline src/System.Threading.Overlapped/tests/ThreadPoolBoundHandle_GetNativeOverlappedStateTests.cs]: Right, the current implementation doesn't treat IAsyncResult specially.  I think there's value in keeping this test case, though, as it seems probable that we could return to the special-casing behavior in the future.  Maybe the comment should change to reflect that.

File: src/System.Threading.Overlapped/src/System/Threading/ClrThreadPoolBoundHandle.cs (hunk 3 above)
Comment (ericeil) [inline src/System.Threading.Overlapped/src/System/Threading/ClrThreadPoolBoundHandle.cs]: I can't believe I didn't add a comment around this....  IIRC, the NullReferenceException happens if the NativeOverlapped is non-null, but has already been free'd.  You can't check for that without calling Unpack (or Free).  This is a case of a usage error (use after free).

**Comment by FiveTimesTheFun** [comment]
The test failure is because the test assumes a file write on an overlapped handle will always execute asynchronously.  I've fixed the test code and will update the PR soon.

File: src/System.Threading.Overlapped/src/System/Threading/DeferredDisposableLifetime.cs (hunk 7 above)
Comment (FiveTimesTheFun) [inline src/System.Threading.Overlapped/src/System/Threading/DeferredDisposableLifetime.cs]: I now dispose the PreAllocatedOverlapped instances, which I believe causes this code path to run.

**Comment by FiveTimesTheFun** [comment]
I believe that addresses the comments.  Please let me know if there are any more before I merge this in under the authority of stoub's "otherwise LGTM"

**Comment by stephentoub** [comment]
I left one more comment that I think needs to be addressed (let me know if there's a reason it's actually ok and I'm just missing it).  After that's addressed, go ahead and merge, though I'd recommend squashing all of your "addressed PR feedback" commits into a single one (separate from the initial dotnet-bot commit).

**Comment by FiveTimesTheFun** [comment]
Thanks Stephen - addressed and pushed.  Will merge after this CI build.

**Pull request merged**